pkg/collector/corechecks/snmp/checkconfig/config.go

```diff
@@ -0,0 +1,482 @@
+package checkconfig
+
+import (
+	"fmt"
+	"path/filepath"
+	"sort"
+	"strings"
+	"time"
+
+	"gopkg.in/yaml.v2"
+
+	"github.com/DataDog/datadog-agent/pkg/autodiscovery/integration"
+	"github.com/DataDog/datadog-agent/pkg/collector/check/defaults"
+	"github.com/DataDog/datadog-agent/pkg/util/log"
+
+	"github.com/DataDog/datadog-agent/pkg/collector/corechecks/snmp/common"
+	"github.com/DataDog/datadog-agent/pkg/collector/corechecks/snmp/metadata"
+)
+
+// Using high oid batch size might lead to snmp calls timing out.
+// For some devices, the default oid_batch_size of 5 might be high (leads to timeouts),
+// and require manual setting oid_batch_size to a lower value.
+var defaultOidBatchSize = 5
+
+var defaultPort = uint16(161)
+var defaultRetries = 3
+var defaultTimeout = 2
+var subnetTagPrefix = "autodiscovery_subnet"
+
+// DefaultBulkMaxRepetitions is the default max rep
+// Using too high max repetitions might lead to tooBig SNMP error messages.
+// - Java SNMP and gosnmp (gosnmp.defaultMaxRepetitions) uses 50
+// - snmp-net uses 10
+const DefaultBulkMaxRepetitions = uint32(10)
+
+// InitConfig maps to a check init config
+type InitConfig struct {
+	Profiles              profileConfigMap `yaml:"profiles"`
+	GlobalMetrics         []MetricsConfig  `yaml:"global_metrics"`
+	OidBatchSize          Number           `yaml:"oid_batch_size"`
+	BulkMaxRepetitions    Number           `yaml:"bulk_max_repetitions"`
+	CollectDeviceMetadata Boolean          `yaml:"collect_device_metadata"`
+	MinCollectionInterval int              `yaml:"min_collection_interval"`
+}
+
+// InstanceConfig maps to a check instance config
+type InstanceConfig struct {
+	IPAddress             string            `yaml:"ip_address"`
+	Port                  Number            `yaml:"port"`
+	CommunityString       string            `yaml:"community_string"`
+	SnmpVersion           string            `yaml:"snmp_version"`
+	Timeout               Number            `yaml:"timeout"`
+	Retries               Number            `yaml:"retries"`
+	OidBatchSize          Number            `yaml:"oid_batch_size"`
+	BulkMaxRepetitions    Number            `yaml:"bulk_max_repetitions"`
+	User                  string            `yaml:"user"`
+	AuthProtocol          string            `yaml:"authProtocol"`
+	AuthKey               string            `yaml:"authKey"`
+	PrivProtocol          string            `yaml:"privProtocol"`
+	PrivKey               string            `yaml:"privKey"`
+	ContextName           string            `yaml:"context_name"`
+	Metrics               []MetricsConfig   `yaml:"metrics"`
+	MetricTags            []MetricTagConfig `yaml:"metric_tags"`
+	Profile               string            `yaml:"profile"`
+	UseGlobalMetrics      bool              `yaml:"use_global_metrics"`
+	ExtraTags             string            `yaml:"extra_tags"` // comma separated tags
+	Tags                  []string          `yaml:"tags"`       // used for device metadata
```

Comment:
1/ `MetricTags` is about:

```
    ## @param metric_tags - list of mappings - optional
    ## Specify tags that you want applied to all metrics. A tag can be applied from a symbol or an OID.
    ## Regular expressions can be used to separate the resulting value into several tags, or get a
    ## substring using the regular Python engine: https://docs.python.org/3/library/re.html
    #
    # metric_tags:
    #   - OID: 1.3.6.1.2.1.1.5.0
    #     symbol: sysName
    #     tag: snmp_host
    #   - OID: 1.3.6.1.2.1.1.5.0
    #     symbol: sysName
    #     match: (.*)-(.*)
    #     tags:
    #       host: \2
    #       device_type: \1
```

https://github.com/DataDog/integrations-core/blob/master/snmp/datadog_checks/snmp/data/conf.yaml.example#L232-L246

It's similar to metrics_tags in profiles, but declared in config file.

2/ `Tags` or `instance.tags` are just static tags form instance. If the same field for other integrations too.
Normally, the Agent will enrich metrics with the metrics with those tags, but we need to deserialise here because we need them for NDM metadata.

3/ ExtraTags is a workaround to pass tags from snmp listener to snmp integration via AD template (see [auto_conf.yaml](https://github.com/DataDog/datadog-agent/blob/main/cmd/agent/dist/conf.d/snmp.d/auto_conf.yaml)) that only works with strings (that's why we need comma separated values), it doesn't accept list of tags for example.

But it wasn't a good design decision since this can be done using another method  described in II-473 . So, once this is done, we can deprecate instance.extra_tags field.

Reply:
Thank you. Could you add context in this struct definition?

Reply:
Sure, added more details here 31214a5a8